src/main/java/com/scalyr/integrations/kafka/AddEventsClient.java

```diff
@@ -74,10 +73,10 @@
   // Separate logger which is used for logging the whole payload when the payload exceeds the limit
   private static final Logger payloadLogger = LoggerFactory.getLogger("com.scalyr.integrations.kafka.eventpayload");
 
-  private final CloseableHttpClient client = HttpClients.createDefault();
+  private static final AsyncHttpClient asyncHttpClient = asyncHttpClient();
   private final ObjectMapper objectMapper = new ObjectMapper();
-  private final ExecutorService senderThread = Executors.newSingleThreadExecutor();
```

Comment:
This question is more due to my limited understanding of async patterns in Java - does it still make sense to use two threads here when using async http client?

In other languages, it depends, but in most cases when you are using async code you don't want to use multiple threads in such scenario, but a single one (this way you actually utilize resources efficiently and you can do that since async code won't block other async code during blocking / waiting operations).

I imagine not using two threads here may also require more code to be refactored and rely on some kind of other "async native" primitive for sleep, but dunno.

Reply:
Addressing both the `Uninterruptibles.sleepUninterruptibly(timeMs, TimeUnit.MILLISECONDS)` and this comment together since they are related.

For `ListenableFuture.addListener(Runnable listener, @Nullable Executor exec)`, if `exec` is null, the callback is executed on the Netty thread, and we should not block or tie up this thread.
https://github.com/AsyncHttpClient/async-http-client#setting-callbacks-on-the-listenablefuture

Since the retry has a delay, we don't want to tie up the Netty thread when `Uninterruptibles.sleepUninterruptibly`sleeps on retry.  That's where this thread pool comes in.  It is possible to use a `ScheduledExecutorService` instead.  There are some advantages to this when we have multiple tasks running in the worker node and they can use the same single `ScheduledExecutorService` thread for the delay instead of 1 thread per task.  This would require more code refactoring to go this route.  It is also possible o use the RxJava library to do an "async native sleep".

I'll look into this to see if there's something we can do without a lot of code refactoring.